Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts

```diff
@@ -0,0 +1,779 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+
+//  BocListe.ts contains client side scripts used by BocList.
+
+class BocList_SelectedRows
+{
+  public Selection: Dictionary<HTMLElement> | number;
+  public Length: number;
+  public Rows: Dictionary<BocList_RowBlock | null>;
+  public SelectRowSelectorControls: JQuery | null;
+  public SelectAllSelectorControls: JQuery | null;
+  public DataRowCount: number;
+
+  constructor(selection: Dictionary<HTMLElement> | number)
+  {
+    this.Selection = selection;
+    this.Length = 0;
+    this.Rows = {}
+    this.SelectRowSelectorControls = null;
+    this.SelectAllSelectorControls = null;
+    this.DataRowCount = 0;
+  }
+
+  public Clear(): void
+  {
+    this.Length = 0;
+    this.Rows = {}
+  }
+
+  OnSelectionChanged(boclist: HTMLElement, flag: boolean): void
+  {
+  }
+}
+
+class BocList_RowBlock
+{
+  public Row: HTMLElement;
+  public SelectorControl: JQuery;
+
+  constructor(row: HTMLElement, selectorControl: JQuery)
+  {
+    if (row.nodeName !== 'TR')
+      throw 'Unexpected element type: \'' + row.nodeName + '\'';
+
+    this.Row = row;
+    this.SelectorControl = selectorControl
+  }
+}
+
+class BocList
+{
+  //  A flag that indicates that the OnClick event for an anchor tag (command) has been raised
+  //  prior to the row's OnClick event.
+  public static _isCommandClick = false;
+
+  //  A flag that indicates that the OnClick event for a selection selectorControl has been raised
+  //  prior to the row's OnClick event.
+  public static _isRowSelectorClickExecuting = false;
+
+  public static _rowSelectionUndefined = -1;
+  public static _rowSelectionDisabled = 0;
+
+  public static _selectedRows: Dictionary<BocList_SelectedRows>;
+
+  //  The css class used for rows in their selected state.
+  public static TrClassNameSelected = 'selected';
+
+  public static _rowSelectionSingleCheckBox = 1;
+  public static _rowSelectionSingleRadioButton = 2;
+  public static rowSelectionMultiple = 3;
```

Comment:
I made them public on purpose because they were `var`s without a scope before, meaning they were globally accessible. Is that incorrect?

Reply:
They were publicly available, but actually private APIs. So we can move them around. Since we already moved them into the class any code that depends on them is broken anyway. `public` vs. `private` has no effect on the resulting JavaScript. `public` and `private` are TypeScript features that cause compile time exceptions. Any `private` member is still publicly available in JavaScript land